### ISSUE TITLE
Seed Ratio is being tracked from torrent size, not downloaded data

### ISSUE THREAD

**Issue by nx6** [transmission/transmission#6087]
### What is the issue?

Recently upgraded from 3.0.0 to 4.0.4 and I'm noticing the seed ratio is now being calculated from the total size of the downloaded torrent (or portion of it) and not the amount of data I actually downloaded from the swarm. So things are being seeded much longer than needed for a specified ratio if I had a torrent with some of the contents downloaded from other sources.

### Which application of Transmission?

transmission-daemon

### Which version of Transmission?

4.0.4

**Comment by tearfur** [comment]
I don't have any opinion about this, I'm just chiming in to clear up some confusion.

> I would like the ratios to be calculated (at least as an option) as uploaded/torrent size

If you are talking about *per-torrent* ratio, in `4.0.5`, the formula is pretty much what you want: `uploaded/(torrent size that you've selected for download + other portions that you have)`

https://github.com/transmission/transmission/blob/a6fe2a64aa7eca089f96006cf082a12f0cde937f/libtransmission/torrent.cc#L1502

In `3.00`, it is calculated as `uploaded/downloaded`, or if `downloaded == 0`, then it is calculated as `uploaded/torrent size that you have` instead.

https://github.com/transmission/transmission/blob/bb6b5a062ee594dfd4b7a12a6b6e860c43849bfd/libtransmission/torrent.c#L1424

Related:
- #6441
- #2873

**Comment by nx6** [comment]
Yeah, So I normally do a seed ratio of 1.1 for most torrents, and now I'm seeing stuff like this:

![app screenshot showing seed ratio for one torrent](https://i.imgur.com/2n5oCRj.png)

**Comment by tearfur** [comment]
I assume this screenshot is from the session stats, and not per-torrent stats? The ratio in session stats is calculated as `uploaded/downloaded` (at least in the WebUI), so it is much larger than `1.1` since the per-torrent ratios are calculated differently.

**Comment by tearfur** [comment]
Since there are opposing opinons here, please discuss and reach a consensus in order for any coding work to happen.

I think it is better if we unify the ratio calculation between session stats and per-torrent stats. However, personally I don't care much which formula we choose in the end.

**Comment by nx6** [comment]
> I assume this screenshot is from the session stats, and not per-torrent stats?

No, that's for one torrent. I had downloaded and seeded a previous version of the torrent. Then a new torrent came out adding a additional files. The local data check found the files from the previous release, so they were counted. It only needed to download 216.4 MB to collect the missing files and finish the torrent after that, which it did, making the total 2.0 GB for the final size. It then seeded 2.2 GB to make a ratio of 1.10 even though it only download a tenth of that.

**Comment by josephj11** [comment]
From what has been discussed so far, there doesn't seem to be one "right" answer. Maybe a global option could be added to allow the user to select the desired ratio calculation. Since any user is unlikely to change this setting very often, this could be implemented as a setting in a parameter file. It wouldn't need a GUI settings dialog component.

**Comment by YURIDUTY** [comment]
> ### What is the issue?
> Recently upgraded from 3.0.0 to 4.0.4 and I'm noticing the seed ratio is now being calculated from the total size of the downloaded torrent (or portion of it) and not the amount of data I actually downloaded from the swarm. So things are being seeded much longer than needed for a specified ratio if I had a torrent with some of the contents downloaded from other sources.
> 
> ### Which application of Transmission?
> transmission-daemon
> 
> ### Which version of Transmission?
> 4.0.4

HI my downloading has taken a pause, but the torrent says it supposed to have more than 16 seeders.
it is like to most of the older torrents I try to download.
for new torrents uploaded in the website this week, I have no problems downloading them super fast.
why is my older torrents not downloading? does my ratio of .03 have anything to do with it? why they won't allow me to download?
please help
thank you

**Comment by josephj11** [comment]
@YURIDUTY  I have no idea what's going on internally, but this is common in older torrents. The fact that seeders exist doesn't mean that they have noticed your client or that they are really seeding anything and are not just sitting there, maybe waiting an hour or more before they look for new peers again. This would be likely to occur when using any torrent client and most likely isn't Transmission's fault.

If you are using a VPN or haven't correctly configured port forwarding on your network, you may be considerably less visible to other peers than you might expect. Only a few VPNs support port forwarding and it will definitely not work by default on Linux (even with no VPN) unless you explicitly configure it. There are lots of threads on how to do this and there are websites you can access which will test if your port is working.

**Comment by tearfur** [comment]
Why the ratio calculation was changed from `3.00` to `4.0.0`:
- #2424
- #2770

**Comment by josephj11** [comment]
I'm a low weight user, but, aside from some useful clarification of the issue, it doesn't appear to be resolved, so I would only suggest closing it if the status is determined to be won't fix.

**Comment by Anusien** [comment]
tearfur mentioned there's not a consensus here, so I'd like to offer a middle ground: calculate the denominator of the seeding ratio as (amount downloaded already + amount remaining in selection to download).

- When the selection of files is fully downloaded, this will give you (uploaded)/(downloaded), which I believe is objectively correct.
- When you haven't done anything weird (updated local files with something from another source, changed the selection you're downloading, etc), you'll get (uploaded)/(size of selection to download), which is the current behavior. (Technically the behavior will slightly change for something like bad blocks.)
- But I believe it will get something intuitive for those weird scenarios.

**Comment by Anusien** [comment]
How do we feel about this approach @tearfur?
